### PR TITLE
utils: allow lxc-attach to set uid / gid

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
//Cc @Disassembler0

**Comment by Disassembler0** [comment]
Works like a charm. Thanks!

File: src/lxc/attach.c [hunk 1]
```diff
@@ -836,25 +838,29 @@ static int attach_child_main(struct attach_clone_payload *payload)
...
+		 * init was started with.
+		 */
+		if (ns_root_uid == LXC_INVALID_UID)
+			lxc_attach_get_init_uidgid(&ns_root_uid, &ns_root_gid);
 
-	if (options->gid != (gid_t)-1)
-		new_gid = options->gid;
+		if (ns_root_uid == LXC_INVALID_UID) {
+			errno = EINVAL;
+			SYSERROR("Failed to find valid namespace uid mapping");
```
Comment (Blub) [inline src/lxc/attach.c]: This function's error case is an `_exit()`, so I don't think we need the `errno`+`SYSERROR` combo.

File: src/lxc/attach.c [hunk 2]
```diff
@@ -836,25 +838,29 @@ static int attach_child_main(struct attach_clone_payload *payload)
...
+			goto on_error;
+		}
+
+		if (ns_root_gid == LXC_INVALID_GID)
+			ns_root_gid = LXC_INVALID_GID;
```
Comment (Blub) [inline src/lxc/attach.c]: This looks like a leftover? (`if (a == b) a = b;`)

File: src/lxc/tools/lxc_attach.c [hunk 3]
```diff
@@ -333,6 +345,14 @@ int main(int argc, char *argv[])
 			goto out;
 	}
 
+	if (my_args.uid) {
```
Comment (Blub) [inline src/lxc/tools/lxc_attach.c]: Unnecessary braces here and in the gid case.

File: src/lxc/start.c [hunk 4]
```diff
@@ -1117,12 +1119,11 @@ static int do_start(void *data)
...
-				  ? 0
-				  : handler->conf->init_uid;
-		gid_t nsgid = (handler->conf->root_nsgid_map != NULL)
-				  ? 0
-				  : handler->conf->init_gid;
+		if (!handler->conf->root_nsuid_map)
+			nsuid = handler->conf->init_uid;
+
+		if (!handler->conf->root_nsgid_map)
+			nsgid = handler->conf->init_uid;
```
Comment (Blub) [inline src/lxc/start.c]: `init_uid` -> `init_gid`

File: src/lxc/attach.c (hunk 1 above)
Comment (brauner) [reply src/lxc/attach.c]: done

File: src/lxc/tools/lxc_attach.c (hunk 3 above)
Comment (brauner) [reply src/lxc/tools/lxc_attach.c]: done

File: src/lxc/start.c (hunk 4 above)
Comment (brauner) [reply src/lxc/start.c]: done

**Comment by brauner** [comment]
@Blub, updated.

**Comment by brauner** [comment]
jenkins: test this please

**Pull request merged**